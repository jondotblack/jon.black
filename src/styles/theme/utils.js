import { baseFontSize } from './constants';

/**
 * Internal helper function
 * Convert pixel value to relative unit
 * @param {string} unit
 * @param {number} base
 * @returns {function(number): string}
 */
export function pxToRelUnit(unit, base) {
  return (x) => `${parseFloat(x) / base}${unit}`;
}

// Relative unit functions
export const pxToEm = pxToRelUnit('em', baseFontSize);
export const pxToRem = pxToRelUnit('rem', baseFontSize);
export const mapToEm = (arr) => arr.map(pxToEm);
export const mapToRem = (arr) => arr.map(pxToRem);

/**
 * Maps breakpoints dictionary, and creates utilities for
 * constructing media queries with those keys.
 *
 * This function takes an object structured
 * like so:
 * ```
 * {
 *   sm: 640px,
 *   md: 960px,
 *   lg: 1280px
 * }
 *
 * The mapping function will be applied to each of the values, and the keys
 * will be used to generate the correct media query string.
 *
 * For example, `media.lessThan('sm')` will produce a string like
 * `(max-width: 640px)` that can be used for styled components.
 *
 * @param {object} breakpoints - the breakpoints object
 */
export function mapMedia(breakpoints) {
  const mapQuery = (func, keys, ...args) => {
    const values = args.flat().map(func);
    const queries = keys.map((key, i) => `(${key}: ${values[i]})`);
    return queries.join(' and ');
  };

  const breakpoint = (key) => breakpoints[key];
  const [min, max] = ['min-width', 'max-width'];

  return {
    // Returns breakpoint value
    breakpoint,
    // Returns true if breakpoint is less than key, i.e. lessThan('sm')
    lessThan: (key) => mapQuery(breakpoint, [max], key),
    // Returns true if breakpoint is more than key, i.e. moreThan('sm')
    moreThan: (key) => mapQuery(breakpoint, [min], key),
    // Returns true if breakpoint is between keys, i.e. between('sm', 'lg')
    between: (...keys) => mapQuery(breakpoint, [min, max], ...keys),
  };
}
