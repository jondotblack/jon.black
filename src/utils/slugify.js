import deburr from 'lodash/deburr';

/* eslint-disable no-useless-escape */
export const slugify = (string) =>
  deburr(string)
    .trim()
    .toLowerCase()
    .replace(/\s+/g, '-') // replace spaces with -
    .replace(/&/g, 'and') // replace & with 'and'
    .replace(/\./g, '-') // replace . with -
    .replace(/([^a-zA-Z0-9\._-]+)/, '');
/* eslint-enable */
