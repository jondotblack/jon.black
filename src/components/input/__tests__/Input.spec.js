import React from 'react';
import renderer from 'react-test-renderer';
import { act, fireEvent, render, screen } from 'utils/testing';

import { Input } from '../index';

const mockPlaceholder = 'george harrison';
const mockValue = 'john lennon';

describe('Input', () => {
  it('should render correctly with all props', () => {
    const tree = renderer.create(<Input name="name" placeholder={mockPlaceholder} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should update value correctly', async () => {
    render(<Input name="name" placeholder={mockPlaceholder} />);

    const name = screen.getByPlaceholderText(mockPlaceholder);

    await act(async () => {
      await fireEvent.change(name, { target: { value: mockValue } });
    });

    expect(name.value).toEqual(mockValue);
  });
});
