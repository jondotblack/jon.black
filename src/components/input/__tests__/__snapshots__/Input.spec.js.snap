// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly with all props 1`] = `
.emotion-0 {
  font-size: 18px;
  line-height: 24px;
  color: #40413b;
  background: transparent;
  border: 0;
}

.emotion-0:focus {
  outline: none;
}

<input
  aria-label="name"
  className="emotion-0 emotion-1"
  name="name"
  placeholder="george harrison"
  type="text"
/>
`;
